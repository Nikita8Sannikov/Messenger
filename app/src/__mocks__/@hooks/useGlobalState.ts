export const mockUseAppSelector = jest.fn();
export const mockUseAppDispatch = jest.fn();