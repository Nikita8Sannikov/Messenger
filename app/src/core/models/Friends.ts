import { type FriendsTab } from "common-types";

import type FriendsController from "@core/controllers/FriendsController";
import { type IUser } from "@custom-types/models.types";

// Контракт модели "Сервис друзей"
export interface Friends {
    friendsController: FriendsController;
    
    checkOnlineUsers(users: IUser[]): void;
    removeOnlineUser(userId: string): void;
    getFriends(type: FriendsTab): void;
    loadMore(type: FriendsTab, resolve: () => void): void;
    search(type: FriendsTab, value: string): void;

    followFriend(friendId: string): void;
    writeMessage(friendId: string): void;
    addFriend(friendId: string): void;
    accept(friendId: string): void;
    leftInFollowers(friendId: string): void;
    unfollow(friendId: string): void;
    deleteFriend(friendId: string): void;
    blockFriend(friendId: string): void;
    unblock(friendId: string): void;
};