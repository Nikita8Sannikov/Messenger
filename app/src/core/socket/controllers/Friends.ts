import EventEmitter from "eventemitter3";

import Logger from "@service/Logger";
import { validateHandleEvent } from "@core/socket/validation";
import { addFriend, deleteFriend } from "@store/friend/slice";
import { setFriendNotification } from "@store/main/slice";
import { FriendsNoticeTypes, SocketActions } from "@custom-types/enums";
import { AppDispatch } from "@custom-types/redux.types";
import { SocketType } from "@custom-types/socket.types";
import { SocketEvents } from "@custom-types/events";

const logger = Logger.init("Socket:FriendsController");

// Контроллер по управлению сокет событиями друзей
export default class FriendsController extends EventEmitter {
    constructor(private readonly _socket: SocketType, private readonly _dispatch: AppDispatch) {
        super();

        this._init();
    }

    private _init() {
        // Подписываемся на пользователя
        this._socket.on(SocketActions.ADD_TO_FRIENDS, (callback) => {
            const validateData = validateHandleEvent(SocketActions.ADD_TO_FRIENDS, {});
            
            if (validateData.success) {
                logger.debug("SocketActions.ADD_TO_FRIENDS");
                this._dispatch(setFriendNotification(FriendsNoticeTypes.ADD));
            }

            this.emit(SocketEvents.SEND_ACK, validateData, callback);
        });

        // Пользователь добавил меня в друзья после моей заявки
        this._socket.on(SocketActions.ACCEPT_FRIEND, ({ user }, callback) => {
            const validateData = validateHandleEvent(SocketActions.ACCEPT_FRIEND, { user });
            
            if (validateData.success) {
                logger.debug(`SocketActions.ACCEPT_FRIEND [userId=${user.id}]`);
                this._dispatch(addFriend(user));
            }

            this.emit(SocketEvents.SEND_ACK, validateData, callback);
        });

        // Отписываемся от пользователя
        this._socket.on(SocketActions.UNSUBSCRIBE, (callback) => {
            const validateData = validateHandleEvent(SocketActions.UNSUBSCRIBE, {});

            if (validateData.success) {
                logger.debug("SocketActions.UNSUBSCRIBE");
                this._dispatch(setFriendNotification(FriendsNoticeTypes.REMOVE));
            }
            
            this.emit(SocketEvents.SEND_ACK, validateData, callback);
        });

        // Пользователь заблокировал меня
        this._socket.on(SocketActions.BLOCK_FRIEND, ({ userId }, callback) => {
            const validateData = validateHandleEvent(SocketActions.BLOCK_FRIEND, { userId });
            
            if (validateData.success) {
                logger.debug(`SocketActions.BLOCK_FRIEND [userId=${userId}]`);
                this._dispatch(deleteFriend(userId));
            }

            this.emit(SocketEvents.SEND_ACK, validateData, callback);
        });
    }
}