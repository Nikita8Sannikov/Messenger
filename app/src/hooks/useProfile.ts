import useMainClient from "@hooks/useMainClient";

// Возврат сущности "Профиль" конкретного пользователя по переданному идентификатору
export default function useProfile(userId?: string) {
	const mainClient = useMainClient();
	return mainClient.getProfile(userId);
}
