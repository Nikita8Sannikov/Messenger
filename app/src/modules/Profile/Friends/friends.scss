@import "../../../styles/abstract/colors.scss";

.friends-container {
    position: relative;
    min-height: 100px;
    color: $black;

    &__top-friends {
        &__spinner {
            padding: 15px;
            width: 100%;
            display: flex;
            justify-content: center;
            align-items: center;
        }

        &__list {
            width: 100%;
            display: flex;
            flex-direction: row;
            flex-wrap: wrap;
        }

        &__item {
            width: 33%;
            display: flex;
            flex-direction: column;
            align-items: center;
            justify-content: center;
            padding: 0;
            font-size: 12px;
            font-weight: 400;

            &:hover {
                cursor: pointer;
                background: $very-medium-grey;
                border-radius: 6px;
            }
        }
    }
}