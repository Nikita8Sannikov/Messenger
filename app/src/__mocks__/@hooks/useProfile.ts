import { type IPhoto } from "@custom-types/models.types";

const mockNotificationSettingsService = {
	soundEnabled: false,
	messageSound: false,
	friendRequestSound: false,
	changeSoundNotifications: jest.fn(),
};

const mockPhotosService = {
	photos: [] as IPhoto[],
	count: 0,
	getAllPhotos: jest.fn(),
	onClickPhoto: jest.fn(),
	addPhotoFromAvatar: jest.fn(),
	addPhotos: jest.fn(),
	deletePhoto: jest.fn(),
};

const mockUserDetailsService = {
	updateDetails: jest.fn(),
	syncUserDetails: jest.fn(),
	getFriendsText: jest.fn(),
	getSubscribersText: jest.fn(),
	getPhotosText: jest.fn(),
	getAudiosText: jest.fn(),
	getVideosText: jest.fn(),
};

const mockUserService = {
	photosService: mockPhotosService,
	detailsService: mockUserDetailsService,
	settingsService: mockNotificationSettingsService,
	id: "",
	firstName: "",
	thirdName: "",
	phone: "",
	email: "",
	fullName: "",
	avatarUrl: "",
	avatarCreateDate: "",
	updateUser: jest.fn(),
	syncInfo: jest.fn(),
	changeAvatar: jest.fn(),
};

const mockProfileService = {
	isMe: false,
	userService: mockUserService,
	photosService: mockPhotosService,
	deleteAccount: jest.fn(),
	onSetAvatar: jest.fn(),
	onDeleteAvatar: jest.fn(),
	onClickAvatar: jest.fn(),
	getFriends: jest.fn(),
	editInfo: jest.fn(),
};

const useProfile = jest.fn().mockReturnValue(mockProfileService);

export default useProfile;
export { 
	mockProfileService, 
	mockUserService, 
	mockPhotosService, 
	mockUserDetailsService, 
	mockNotificationSettingsService, 
};