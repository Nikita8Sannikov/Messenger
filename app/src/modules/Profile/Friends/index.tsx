import { useState, useEffect, memo } from "react";
import { useTranslation } from "react-i18next";

import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";
import ListItemAvatar from "@mui/material/ListItemAvatar";

import GridComponent from "@components/ui/Grid";
import PaperComponent from "@components/ui/Paper";
import AvatarComponent from "@components/ui/avatar";
import NoDataComponent from "@components/ui/no-data";
import SpinnerComponent from "@components/ui/spinner";
import { FriendsTab, MainFriendTabs, Pages } from "@custom-types/enums";
import { IUser } from "@custom-types/models.types";
import { useAppSelector } from "@hooks/useGlobalState";
import useProfile from "@hooks/useProfile";
import { selectMainState } from "@store/main/slice";
import { selectFriendState } from "@store/friend/slice";
import { getFullName } from "@utils/index";

import "./friends.scss";

export type onClickBlockType = (
    pathname: Pages,
    query: {
        mainTab: MainFriendTabs;
        tab?: FriendsTab;
    }
) => void;

interface IFriends {
    onlineFriends?: boolean;
    onClickBlock: onClickBlockType;
};

interface IFriendsState {
    title: string;
    count: number | string;
    users: IUser[] | null;
};

export default memo(function Friends({ onlineFriends = false, onClickBlock }: IFriends) {
    const { t } = useTranslation();

    const [loading, setLoading] = useState(false);
    const [friendsState, setFriendsState] = useState<IFriendsState>({
        title: t("profile-module.friends"),
        count: "-",
        users: null
    });

    const { topFriends, friendsCount } = useAppSelector(selectFriendState);
    const { onlineUsers } = useAppSelector(selectMainState);

    const profile = useProfile();

    // Формируем объект данных компонента
    useEffect(() => {
        const state = {
            title: onlineFriends ? t("profile-module.friends_online") : t("profile-module.friends"),
            count: onlineFriends ? onlineUsers.length : friendsCount,
            users: onlineFriends ? onlineUsers.slice(0, 6) : topFriends
        };

        setFriendsState(state);
    }, [onlineFriends, onlineUsers, topFriends, friendsCount]);

    // Получаем количество друзей, подписчиков и подгружаем первые 6 из них
    useEffect(() => {
        if (!onlineFriends) {
            profile.getFriends(setLoading);
        }
    }, []);

    // Обрабока клика по блоку
    const blockHandler = () => {
        const params: { mainTab: MainFriendTabs; tab?: FriendsTab; } = { mainTab: MainFriendTabs.allFriends };

        if (onlineFriends) {
            params.tab = FriendsTab.online;
        }

        onClickBlock(Pages.friends, params);
    };

    const { title, count, users } = friendsState;

    return <GridComponent className="friends-container__grid">
        <PaperComponent className="friends-container paper-block">
            <div className="block-title" onClick={blockHandler}>
                {title} <span className="counter">{count}</span>
            </div>

            {loading
                ? <SpinnerComponent />
                : users && users.length
                    ? <List className="friends-container__top-friends__list">
                        {users.map(user => {
                            const userName = getFullName(user);

                            return <ListItem className="friends-container__top-friends__item" key={user.id}>
                                <ListItemAvatar className="friends-container__top-friends__item__avatar-block">
                                    <AvatarComponent
                                        src={user.avatarUrl}
                                        alt={userName}
                                        isOnline={Boolean(onlineUsers.find(onlineUser => onlineUser.id === user.id))}
                                        size={50}
                                    />
                                </ListItemAvatar>

                                {user.firstName}
                            </ListItem>
                        })}
                    </List>
                    : <NoDataComponent text={t("profile-module.no_data", { isOnline: onlineFriends ? t("profile-module.online") : "" })} />
            }
        </PaperComponent>
    </GridComponent>
});