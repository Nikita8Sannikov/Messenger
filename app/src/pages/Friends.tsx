import { FriendsTab } from "common-types";
import { useEffect, useRef } from "react";
import { useNavigate, useParams } from "react-router-dom";

import GridComponent from "@components/ui/grid";
import PaperComponent from "@components/ui/paper";
import Content from "@modules/friends/content";
import PossibleFriends from "@modules/friends/possible-friends";
import MainTabs from "@modules/friends/tabs";
import { type VirtualListHandle } from "@modules/virtual-list/list";
import useFriendsStore from "@store/friends";
import LinkComponent from "@components/ui/link";

import "@styles/pages/friends.scss";
import useProfile from "@hooks/useProfile";
import BigUserAvatar from "@components/services/avatars/big-user-avatar";
import { Pages } from "@custom-types/enums";
import { goToAnotherProfile } from "@utils/index";
import usePrepareAnotherUser from "@hooks/usePrepareAnotherUser";
import SpinnerComponent from "@components/ui/spinner";
import useUser from "@hooks/useUser";

// Страница "Друзья"
export default function Friends() {
	const virtualRef = useRef<VirtualListHandle | null>(null);

	const { userId } = useParams();
	const navigate = useNavigate();

	const { isLoading: isLoadingAnotherUser, isExistProfile } = usePrepareAnotherUser();

	const mainTab = useFriendsStore(state => state.mainTab);
	const profile = useProfile(isLoadingAnotherUser || !isExistProfile ? undefined : userId);
	const user = useUser(isLoadingAnotherUser || !isExistProfile ? undefined : userId);
	const friendsService = user.friendsService;

	// Получение друзей при загрузке страницы
	useEffect(() => {
		friendsService.getFriends(FriendsTab.MY);
	}, [userId]);

	// Обработка перехода на чужой профиль
	const onClickAnotherUser = () => {
		navigate(goToAnotherProfile(Pages.profile, userId));
	};

	// При первом входе на страницу по урлу другого пользователя, показываем спиннер, пока ждем его подгрузки и создание его профиля
	if (isLoadingAnotherUser || !isExistProfile) {
		return <div className="user-profile-page-spinner">
			<SpinnerComponent />
		</div>;
	}

	return <GridComponent container spacing={2} className="friends-container">
		<GridComponent xs={8} className="friends-container__content">
			<PaperComponent className="paper-block friends-container__content__wrapper">
				<Content virtualRef={virtualRef} />
			</PaperComponent>
		</GridComponent>

		<GridComponent xs={4} direction="column" spacing={2}>
			<GridComponent container spacing={2}>
				<GridComponent>
					<PaperComponent className="paper-block">
						{profile.isMe
							? <MainTabs virtualRef={virtualRef} />
							: <div className="friends-container__another-user">
								<BigUserAvatar
									userId={user.id}
									src={user.avatarUrl}
									alt={user.fullName}
								/>

								<div className="friends-container__another-user__info">
									{user.fullName}

									<LinkComponent
										variant="body2"
										onClick={onClickAnotherUser}
										underline="hover"
										className="friends-container__another-user__info__link"
									>
										Перейти к странице
									</LinkComponent>
								</div>
							</div>
						}
					</PaperComponent>
				</GridComponent>

				{mainTab === FriendsTab.SEARCH || !profile.isMe
					? null
					: <GridComponent className="friends-container__possible-friends">
						<PaperComponent className="paper-block">
							<PossibleFriends virtualRef={virtualRef} />
						</PaperComponent>
					</GridComponent>
				}
			</GridComponent>
		</GridComponent>
	</GridComponent>;
};