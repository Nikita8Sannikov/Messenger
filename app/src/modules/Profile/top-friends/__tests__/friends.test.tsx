import { act, render, screen } from "@testing-library/react";

import useProfile from "@hooks/useProfile";
import Friends from "@modules/profile/top-friends/friends";
import { mockProfileService } from "../../../../__mocks__/@hooks/useProfile";

jest.mock("@hooks/useProfile");
jest.mock("react-router-dom", () => ({
	useNavigate: () => jest.fn(),
}));

describe("Friends", () => {
	beforeEach(() => {
		(useProfile as jest.Mock).mockReturnValue(mockProfileService);
	});

	it("renders with no friends", async () => {
		await act(async () => {
			const { container } = render(<Friends />);
			expect(container).toMatchSnapshot();
		});
		expect(screen.getByTestId("no-data")).toBeInTheDocument();
	});

	it("calls getFriends on mount", async () => {
		await act(async () => {
			render(<Friends />);
		});
		expect(mockProfileService.getFriends).toHaveBeenCalled();
	});

	it("renders loading state", async () => {
		const { container, rerender } = render(<Friends />);
		expect(screen.queryByTestId("spinner")).not.toBeInTheDocument();
        
		const loadingComponent = <Friends />;
		loadingComponent.type.defaultProps = { loading: true };
		await act(async () => {
			rerender(loadingComponent);
		});
		expect(container).toMatchSnapshot();
	});

	it("renders friends count", async () => {
		await act(async () => {
			const { container } = render(<Friends />);
			expect(container).toMatchSnapshot();
		});
		expect(screen.getByText("0")).toBeInTheDocument();
	});

	it("renders friends title", async () => {
		await act(async () => {
			const { container } = render(<Friends />);
			expect(container).toMatchSnapshot();
		});
		expect(screen.getByText("Friends")).toBeInTheDocument();
	});
});
