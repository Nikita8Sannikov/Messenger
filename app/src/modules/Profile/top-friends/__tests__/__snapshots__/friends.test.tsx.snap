// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends renders friends count 1`] = `<div />`;

exports[`Friends renders friends title 1`] = `<div />`;

exports[`Friends renders loading state 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          0
        </span>
      </div>
      <div
        class="opacity-text no-items"
        data-testid="no-data"
      >
        The friends list is empty at the moment.
      </div>
    </div>
  </div>
</div>
`;

exports[`Friends renders with no friends 1`] = `<div />`;
